src/stk/databases/mongo_db/constructed_molecule.py

```diff
@@ -689,38 +689,86 @@ def _get_molecule_keys(self, entry):
                 yield key, value
 
     def get_all(self):
-        for entry in self._constructed_molecules.find():
-            # Do 'or' query over all key value pairs.
-            query = {'$or': [
-                {key: value}
-                for key, value in self._get_molecule_keys(entry)
-            ]}
-
-            molecule_json = self._molecules.find_one(query)
-            if molecule_json is None:
-                raise KeyError(
-                    'No molecule found in the database associated '
-                    f'with a position matrix with query: {query}. '
-                    'This suggests your database is corrupted.'
-                )
-
-            position_matrix = self._position_matrices.find_one(query)
-            if position_matrix is None:
-                raise KeyError(
-                    'No position matrix found in the database '
-                    'associated with a position matrix with query: '
-                    f'{query}. This suggests your database is '
-                    'corrupted.'
-                )
+        # Get all potential indices.
+        pos_mat_indices = self._position_matrices.index_information()
+        molecules_indices = self._molecules.index_information()
+        constmolecules_indices = (
+            self._constructed_molecules.index_information()
+        )
 
-            yield self._dejsonizer.from_json(
-                json={
-                    'molecule': molecule_json,
-                    'constructedMolecule': entry,
-                    'matrix': position_matrix,
-                    'buildingBlocks': tuple(map(
-                        self._get_building_block,
-                        entry['BB'],
-                    ))
-                },
+        keys = tuple(set((
+            val['key'][0][0]
+            for val in (
+                list(pos_mat_indices.values())
+                + list(molecules_indices.values())
+                + list(constmolecules_indices.values())
```

Comment:
how bout

```python
indices = itertools.chain(
    self._positoin_matrices.index_information().values(),
    self.molecules.index_informatoin.values(),
    self._constructed_molecules.index_information.values(),
)
keys = dedupe(
   index['key'][0][0] for index in indices 
)
```

* Uses chain instead of list concatenation - O(1) vs O(N) memory use, + O(N) vs O(N^2) time
* No need to make a tuple at the end, since you are only using keys to iterate though them: O(1) vs O(N) memory + O(1) vs O(N) time.

Reply:
I like it!